presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/SetOperationMerge.java

```diff
@@ -55,58 +59,117 @@ public SetOperationMerge(SetOperationNode node, Context context, SetOperationNod
-        if (sources.stream().noneMatch(node.getClass()::isInstance)) {
-            return Optional.empty();
-        }
-
         ImmutableListMultimap.Builder<Symbol, Symbol> newMappingsBuilder = ImmutableListMultimap.builder();
+        boolean resultIsDistinct = false;
+        boolean rewritten = false;
 
         for (int i = 0; i < sources.size(); i++) {
             PlanNode source = sources.get(i);
-            if (node.getClass().equals(source.getClass())) {
-                SetOperationNode setOperationNode = (SetOperationNode) source;
-                addMergedMappings(setOperationNode, i, newMappingsBuilder);
+
+            // Determine if set operations can be merged and whether the resulting set operation is quantified DISTINCT or ALL
+            Optional<Boolean> mergedQuantifier = mergedQuantifierIsDistinct(node, source);
+            if (mergedQuantifier.isPresent()) {
+                addMergedMappings((SetOperationNode) source, i, newMappingsBuilder);
+                resultIsDistinct |= mergedQuantifier.get();
+                rewritten = true;
             }
             else {
                 // Keep mapping as it is
                 addOriginalMappings(source, i, newMappingsBuilder);
             }
         }
 
-        return Optional.of(instantiator.create(node.getId(), newSources, newMappingsBuilder.build(), node.getOutputSymbols()));
+        if (!rewritten) {
+            return Optional.empty();
+        }
+        if (node instanceof UnionNode) {
+            return Optional.of(new UnionNode(node.getId(), newSources, newMappingsBuilder.build(), node.getOutputSymbols()));
+        }
+        return Optional.of(new IntersectNode(node.getId(), newSources, newMappingsBuilder.build(), node.getOutputSymbols(), resultIsDistinct));
+    }
+
+    /**
+     * Check if node and child are mergable based on their set operation type and quantifier.
+     * <p>
+     * For parent and child of type UNION, merge is always possible and the assumed quantifier is ALL, because UnionNode always represents UNION ALL.
+     * <p>
+     * For parent and child of type INTERSECT, merge is always possible:
+     * - if parent and child are both INTERSECT ALL, the resulting set operation is INTERSECT ALL
+     * - otherwise, the resulting set operation is INTERSECT DISTINCT
+     * <p>
+     * For parent and child of type EXCEPT:
+     * - if parent is EXCEPT DISTINCT and child is EXCEPT ALL, merge is not possible
+     * - if parent and child are both EXCEPT DISTINCT, the resulting set operation is EXCEPT DISTINCT
+     * - if parent and child are both EXCEPT ALL, the resulting set operation is EXCEPT ALL
+     * - if parent is EXCEPT ALL and child is EXCEPT DISTINCT, the resulting set operation is EXCEPT DISTINCT
+     * <p>
+     * Optional.empty() indicates that merge is not possible.
+     */
+    private Optional<Boolean> mergedQuantifierIsDistinct(SetOperationNode node, PlanNode child)
```

Comment:
What child does this refer to? For EXCEPT, the first one vs any of the others matters.

Reply:
For EXCEPT, this is supposed to be called only for the first child. For INTERSECT and UNION -- for all children.
This is responsibility of the caller to call the method for the appropriate children and aggregate the results.